### PR TITLE
Fix win_dacl problems with SIDs

### PR DESCRIPTION
**Pull request by twangboy**
### What does this PR do?
Fixes an issue where `file.manage_file` would crash if trying to manage a file that had a security principal with an unmapped name. When you look at the object using the GUI in Windows, it just shows the SID, so this just returns the SID when there is no mapped security principal.

Fixes the SID for NULL SID (S-1-0-0). It had the SID for Everyone (S-1-1-0)

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/45126

### Tests written?
No

### Commits signed with GPG?
Yes

### CODE REVIEW

**Comment by BenjaminSchiborr** [comment]
@twangboy, I've tried out the referenced PR on our test systems. Unfortunately this does not fix the underlying issue and results in a comparable stack:
```
salt '*MS*' file.manage_file C:/Deployment/im_a_cool_file '' '' '' '' '' '' '' '' ''                                                                                                                     
VADAUTOM41MS001:                                                                                                                                                                                                          
    The minion function caused an exception: Traceback (most recent call last):                                                                                                                                           
      File "c:\salt\bin\lib\site-packages\salt\minion.py", line 1470, in _thread_return                                                                                                                                   
        return_data = executor.execute()                                                                                                                                                                                  
      File "c:\salt\bin\lib\site-packages\salt\executors\direct_call.py", line 28, in execute                                                                                                                             
        return self.func(*self.args, **self.kwargs)                                                                                                                                                                       
      File "c:\salt\bin\lib\site-packages\salt\modules\file.py", line 4906, in manage_file                                                                                                                                
        kwargs.get('win_inheritance'))                                                                                                                                                                                    
      File "c:\salt\bin\lib\site-packages\salt\modules\win_file.py", line 1588, in check_perms                                                                                                                            
        cur_perms = salt.utils.win_dacl.get_permissions(path)                                                                                                                                                             
      File "c:\salt\bin\lib\site-packages\salt\utils\win_dacl.py", line 1513, in get_permissions                                                                                                                          
        return obj_dacl.list_aces()                                                                                                                                                                                       
      File "c:\salt\bin\lib\site-packages\salt\utils\win_dacl.py", line 828, in list_aces                                                                                                                                 
        user, a_type, a_prop, a_perms, inh = self._ace_to_dict(ace)                                                                                                                                                       
      File "c:\salt\bin\lib\site-packages\salt\utils\win_dacl.py", line 871, in _ace_to_dict                                                                                                                              
        principal = get_name(sid)                                                                                                                                                                                         
      File "c:\salt\bin\lib\site-packages\salt\utils\win_dacl.py", line 1138, in get_name                                                                                                                                 
        return win32security.LookupAccountSid(None, sid_obj)[0]                                      
```

I think the reason for that is that a _CommandExecutionError_ is intercepted around _principal = get_name(sid)_. However, the error is a _pywintypes.error_.

 If I intercept this one instead the command succeeds.

**Comment by twangboy** [comment]
@BenjaminSchiborr The `get_name` function intercepts the `pywintypes.error` and returns a `CommandExecutionError`. This should be caught in the `_ace_to_dict` function and just return the SID. Is there anything in the stack trace below what you show above?

**Comment by BenjaminSchiborr** [comment]
@twangboy, no, this was the full stack trace.

I think it only throws a `CommandExecutionError` in `get_name` if the error thrown inside is a `TypeError`. See the following snippet from `get_name`:
```python
    try:
        return win32security.LookupAccountSid(None, sid_obj)[0]
    except TypeError:
        raise CommandExecutionError(
            'Could not find User for {0}'.format(principal))
```
`LookupAccountSid` throws the `pywintypes.error`, which is not intercepted (as it is not a `TypeError`). Therefore it bubbles up to the calling function, which then intercepts just a `CommandExecutionError`.

**Comment by twangboy** [comment]
@BenjaminSchiborr Are you sure you tried the file in the PR (https://github.com/twangboy/salt/blob/35a417f510fbe7fb16a8b350945c1e250499256f/salt/utils/win_dacl.py)? The reason I ask is because the `try/except` block is different. Mine is showing:
```
    try:
        return win32security.LookupAccountSid(None, sid_obj)[0]
    except (pywintypes.error, TypeError) as exc:
        if type(exc) == pywintypes.error:
            win_error = win32api.FormatMessage(exc.winerror).rstrip('\n')
            message = 'Error resolving {0} ({1})'.format(principal, win_error)
        else:
            message = 'Error resolving {0}'.format(principal)

        raise CommandExecutionError(message)
```

**Comment by BenjaminSchiborr** [comment]
@twangboy that makes sense. I just applied the delta to the minions on our test systems. For v2017.7.2 my statement is valid, but this has already changed for 2017.7 branch. Sorry for the confusion and thanks for the quick response! Oh, and a Happy New Year!

**Comment by twangboy** [comment]
@lorengordon Would you mind reviewing this?

File: salt/utils/win_dacl.py [hunk 1]
```diff
@@ -1170,17 +1173,17 @@ def get_owner(obj_name):
...
         # Incorrect function error (Windows Server 2008+)
         if exc.winerror == 1 or exc.winerror == 50:
-            owner_sid = 'S-1-1-0'
+            owner_sid = 'S-1-0-0'
         else:
             raise CommandExecutionError(
                 'Failed to get owner: {0}'.format(exc.strerror))
 
-    return get_name(win32security.ConvertSidToStringSid(owner_sid))
+    return get_name(owner_sid)
```
Comment (lorengordon) [inline salt/utils/win_dacl.py]: @twangboy I think I understand all of the changes except this one. Mind commenting on why this was required?

Comment (twangboy) [reply salt/utils/win_dacl.py]: The get_name function was modified to accept any type of input and return the name. So it already does a `win32security.ConvertSidToStringSid` check. No need to do it twice.

**Review by lorengordon — APPROVED** [review]
Looks good to me!

**Review by rallytime — APPROVED** [review]

**Pull request merged**